kate/recovery/src/com.rs

```diff
@@ -162,39 +115,51 @@ pub fn reconstruct_app_extrinsics(
 /// * `dimensions` - Extended matrix dimensions
 /// * `cells` - Cells from required columns, at least 50% cells per column
 pub fn reconstruct_extrinsics(
-	index: &AppDataIndex,
-	dimensions: &ExtendedMatrixDimensions,
-	cells: Vec<DataCell>,
+	index: &index::AppDataIndex,
+	dimensions: &matrix::Dimensions,
+	cells: Vec<data::DataCell>,
 ) -> Result<Vec<(u32, Vec<Vec<u8>>)>, ReconstructionError> {
 	let data = reconstruct_available(dimensions, cells)?;
 	let ranges = index.data_ranges();
-	unflatten_padded_data(ranges, data, CHUNK_SIZE).map_err(ReconstructionError::DataDecodingError)
+	unflatten_padded_data(ranges, data).map_err(ReconstructionError::DataDecodingError)
 }
 
 fn reconstruct_available(
-	dimensions: &ExtendedMatrixDimensions,
-	cells: Vec<DataCell>,
+	dimensions: &matrix::Dimensions,
+	cells: Vec<data::DataCell>,
 ) -> Result<Vec<u8>, ReconstructionError> {
-	let mut column_numbers: Vec<u16> = vec![];
-	let mut data: Vec<u8> = vec![];
-	let cells_map = map_cells(dimensions, cells)?;
-	for column_number in 0..dimensions.cols as u16 {
-		match cells_map.get(&column_number) {
-			None => data.extend(vec![0; dimensions.rows / 2 * CHUNK_SIZE]),
+	let columns = map_cells(dimensions, cells)?;
+
+	let scalars = (0..dimensions.cols)
+		.map(|col| match columns.get(&col) {
+			None => Ok(vec![None; dimensions.rows as usize]),
 			Some(column_cells) => {
-				if column_cells.len() < dimensions.rows / 2 {
-					return Err(ReconstructionError::InvalidColumn(column_number));
+				if column_cells.len() < dimensions.rows as usize {
+					return Err(ReconstructionError::InvalidColumn(col));
 				}
 				let cells = column_cells.values().cloned().collect::<Vec<_>>();
-				let scalars = reconstruct_column(dimensions.rows, &cells)
-					.map_err(ReconstructionError::ColumnReconstructionError)?;
-				let column_data = scalars.iter().flat_map(|e| e.to_bytes());
-				column_numbers.push(column_number);
-				data.extend(column_data);
+				reconstruct_column(dimensions.extended_rows() as usize, &cells)
+					.map(|scalars| scalars.into_iter().map(Some).collect::<Vec<_>>())
+					.map_err(ReconstructionError::ColumnReconstructionError)
 			},
-		}
+		})
+		.collect::<Result<Vec<Vec<_>>, ReconstructionError>>()?
+		.into_iter()
+		.flatten()
+		.collect::<Vec<_>>();
```

Comment:
QQ: Anyway to avoid this double `collect`?
```
		.collect::<Result<Vec<Vec<_>>, ReconstructionError>>()?
		.into_iter()
		.flatten()
		.collect::<Vec<_>>();
```

Reply:
Its not trivial due to vector nested in result. It could be done imperatively or maybe to use itertools::process_results, or to model things differently (maybe row not assigned to some app id could be none instead of vector of none cells). However, maybe avoiding flattening is OK in our case:
```
		})
		.collect::<Result<Vec<Vec<_>>, ReconstructionError>>()?;

	let mut result: Vec<u8> = Vec::with_capacity(scalars.len() * config::CHUNK_SIZE);

	for (row, col) in dimensions.iter_data() {
		let bytes = scalars
			.get(col)
			.and_then(|col| col.get(row))
			.map(Option::as_ref)
			.unwrap_or(None)
			.map(BlsScalar::to_bytes)
			.unwrap_or_else(|| [0; config::CHUNK_SIZE]);
		result.extend(bytes);
	}
```
wdyt?